[Win] WebInspector: enable HAR export and import

haruhisa-shin opened the pull request:
#### 8cc207109df127720f2ac10fb8d49fd0d78bfed9
<pre>
[Win] WebInspector: enable HAR export and import
<a href="https://bugs.webkit.org/show_bug.cgi?id=253552">https://bugs.webkit.org/show_bug.cgi?id=253552</a>

Reviewed by Fujii Hironori.

Add file open dialog implementations for both Export and Import.
And, canSave() now returns true so that the Export button can be pressed.

* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.h:
* Source/WebKit/UIProcess/Inspector/win/RemoteWebInspectorUIProxyWin.cpp:
(WebKit::RemoteWebInspectorUIProxy::platformSave):
* Source/WebKit/UIProcess/Inspector/win/WebInspectorUIProxyWin.cpp:
(WebKit::systemErrorMessage):
(WebKit::WebInspectorUIProxy::showSavePanelForSingleFile):
(WebKit::WebInspectorUIProxy::platformSave):
* Source/WebKit/UIProcess/win/PageClientImpl.cpp:
(WebKit::PageClientImpl::handleRunOpenPanel):
* Source/WebKit/UIProcess/win/PageClientImpl.h:
* Source/WebKit/WebProcess/Inspector/win/RemoteWebInspectorUIWin.cpp:
(WebKit::RemoteWebInspectorUI::canSave):
* Source/WebKit/WebProcess/Inspector/win/WebInspectorUIWin.cpp:
(WebKit::WebInspectorUI::canSave):

Canonical link: <a href="https://commits.webkit.org/261609@main">https://commits.webkit.org/261609@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/07e74d7db2406eda39ecd9f8c5f2c28ed403eccc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112302 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21439 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/934 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/87/builds/4082 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/120919 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22777 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12497 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/87/builds/4082 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118066 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16926 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/100107 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/105351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98883 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/88/builds/646 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/105351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/13817 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/692 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/105351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14508 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9976 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/19860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52696 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8093 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16309 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebKit/UIProcess/Inspector/win/RemoteWebInspectorUIProxyWin.cpp:
```diff
@@ -36,9 +36,12 @@
 
 namespace WebKit {
 
+using namespace FileSystem;
```
I think you should remove this line becuase most code is using a prefixed style `FileSystem::openFile`. I was asked not to use `using namespace` by reviewers.

haruhisa-shin replied:
Fixed at 19fcde7

fujii commented on Source/WebKit/UIProcess/Inspector/win/RemoteWebInspectorUIProxyWin.cpp:
```diff
@@ -131,9 +134,48 @@ WebPageProxy* RemoteWebInspectorUIProxy::platformCreateFrontendPageAndWindow()
     return m_webView->page();
 }
 
+void RemoteWebInspectorUIProxy::platformSave(Vector<WebCore::InspectorFrontendClient::SaveData>&& saveDatas, bool forceSaveAs)
```
Generally, `UNUSED_PARAM` is used only with `#if`. I think you should do `/* forceSaveAs */` in this case.

haruhisa-shin replied:
Ditto.

fujii commented on Source/WebKit/UIProcess/Inspector/win/RemoteWebInspectorUIProxyWin.cpp:
```diff
@@ -131,9 +134,48 @@ WebPageProxy* RemoteWebInspectorUIProxy::platformCreateFrontendPageAndWindow()
+    Vector<WCHAR, MAX_PATH> filePath(url.path().substring(1).toString().wideCharacters());
+
+    OPENFILENAME ofn;
+    memset(&ofn, 0, sizeof(ofn));
```
> OPENFILENAME ofn { };

Do this to zero fill in C++.

haruhisa-shin replied:
Ditto.

fujii commented on Source/WebKit/UIProcess/Inspector/win/RemoteWebInspectorUIProxyWin.cpp:
```diff
@@ -131,9 +134,48 @@ WebPageProxy* RemoteWebInspectorUIProxy::platformCreateFrontendPageAndWindow()
+
+    // Remove custom URI schemes such as "web-inspector://" and also remove the leading "/".
+    URL url { saveDatas[0].url };
+    Vector<WCHAR, MAX_PATH> filePath(url.path().substring(1).toString().wideCharacters());
```
This is not a right code even though it doesn't crash. The second template argument `MAX_PATH` is a capacity, not the vector size. You should explicitly allocate.

> Vector<WCHAR> filePath(MAX_PATH);
> ...

or

> auto filePath = url.path().substring(1).toString().wideCharacters();
> filePath.grow(MAX_PATH);
> ...

haruhisa-shin replied:
It seems that the `m_inlineBuffer` inside the WTF::Vector is allocating the capacity size, so it did not crash.
However, it is not a good code as you said, so I will fix it.

haruhisa-shin replied:
Fixed at https://github.com/WebKit/WebKit/commit/19fcde7aadadcc06e0da2dd02f27c39ad4e61521
I used `filePath.grow`.

fujii commented on Source/WebKit/UIProcess/Inspector/win/RemoteWebInspectorUIProxyWin.cpp:
```diff
@@ -131,9 +134,48 @@ WebPageProxy* RemoteWebInspectorUIProxy::platformCreateFrontendPageAndWindow()
+    UNUSED_PARAM(forceSaveAs);
+
+    // Remove custom URI schemes such as "web-inspector://" and also remove the leading "/".
+    URL url { saveDatas[0].url };
```
Use `auto`. `auto url = saveDatas[0].url;`

haruhisa-shin replied:
The `savedatas[0].url` is String.
I want to convert it to a URL, so I would change it as follows:
> auto url = URL(saveDatas[0].url);

cdumez replied:
I don't think `auto url = URL(saveDatas[0].url);` is really better than `URL url { saveDatas[0].url };`. I'd recommend not changing it.

haruhisa-shin replied:
No change.

fujii commented on Source/WebKit/UIProcess/Inspector/win/RemoteWebInspectorUIProxyWin.cpp:
```diff
@@ -131,9 +134,48 @@ WebPageProxy* RemoteWebInspectorUIProxy::platformCreateFrontendPageAndWindow()
+
+        String content { saveDatas[0].content };
+        auto contentSize = content.utf8().length();
+        auto bytesWritten = writeToFile(fd, content.utf8().data(), contentSize);
```
`content.utf8()` allocates a buffer. Don't do it twice.
> auto content = saveDatas[0].content.utf8().

haruhisa-shin replied:
Fixed at 19fcde7

fujii commented on Source/WebKit/UIProcess/Inspector/win/RemoteWebInspectorUIProxyWin.cpp:
```diff
@@ -131,9 +134,48 @@ WebPageProxy* RemoteWebInspectorUIProxy::platformCreateFrontendPageAndWindow()
+        auto bytesWritten = writeToFile(fd, content.utf8().data(), contentSize);
+        if (bytesWritten == -1 || bytesWritten != contentSize) {
+            auto errMsg = makeString("Error: writeToFile returns ", bytesWritten, ", contentLength = ", content.length());
+            MessageBox(0, errMsg.wideCharacters().data(), L"Export HAR", MB_OK);
```
Let's add `MB_ICONEXCLAMATION` for error message boxes. `MB_OK | MB_ICONEXCLAMATION`

haruhisa-shin replied:
Ditto.

fujii commented on Source/WebKit/UIProcess/Inspector/win/WebInspectorUIProxyWin.cpp:
```diff
@@ -405,9 +408,44 @@ void WebInspectorUIProxy::platformShowCertificate(const WebCore::CertificateInfo
 }
 
-void WebInspectorUIProxy::platformSave(Vector<WebCore::InspectorFrontendClient::SaveData>&&, bool /* forceSaveAs */)
+void WebInspectorUIProxy::platformSave(Vector<WebCore::InspectorFrontendClient::SaveData>&& saveDatas, bool forceSaveAs)
```
Can you merge some part of this function with above code? Mac port has `WebInspectorUIProxy::showSavePanel`.

haruhisa-shin replied:
I added a new function `WebInspectorUIProxy::showSavePanelForSingleFile` and merged processes.

fujii commented on Source/WebKit/UIProcess/Inspector/win/RemoteWebInspectorUIProxyWin.cpp:
```diff
@@ -131,9 +134,48 @@ WebPageProxy* RemoteWebInspectorUIProxy::platformCreateFrontendPageAndWindow()
+        auto bytesWritten = writeToFile(fd, content.utf8().data(), contentSize);
+        if (bytesWritten == -1 || bytesWritten != contentSize) {
+            auto errMsg = makeString("Error: writeToFile returns ", bytesWritten, ", contentLength = ", content.length());
+            MessageBox(0, errMsg.wideCharacters().data(), L"Export HAR", MB_OK);
```
This is user visible error message. It'd be nice if it's better error message. You can take a text error message. https://stackoverflow.com/q/1387064

haruhisa-shin replied:
I added process to get system error string by `FormatMessageW`.

fujii commented on Source/WebKit/UIProcess/Inspector/win/RemoteWebInspectorUIProxyWin.cpp:
```diff
@@ -131,9 +134,48 @@ WebPageProxy* RemoteWebInspectorUIProxy::platformCreateFrontendPageAndWindow()
+        auto errCode = CommDlgExtendedError();
+        if (errCode) {
+            auto errMsg = makeString("Error: ", errCode);
+            MessageBox(0, errMsg.wideCharacters().data(), L"Export HAR", MB_OK);
```
This is user visible error message. It'd be nice if it's better error message. Especially for `FNERR_INVALIDFILENAME`.

haruhisa-shin replied:
Added error for `FNERR_INVALIDFILENAME`.
Other error codes have not been added, as they are unlikely to be caused by user operation.

The list of errors is based on the following page:
https://learn.microsoft.com/en-us/windows/win32/api/commdlg/nf-commdlg-commdlgextendederror

fujii commented on Source/WebKit/UIProcess/win/PageClientImpl.cpp:
```diff
@@ -232,6 +240,70 @@ void PageClientImpl::preferencesDidChange()
+
+    ofn.lStructSize = sizeof(ofn);
+    ofn.hwndOwner = viewWindow;
+    String allFiles = makeString(allFilesText(), String("\0*.*\0\0", 6));
```
Use `auto` especially if you have a word `String` in both sides.

cdumez replied:
`auto allFiles = makeString(allFilesText(), "\0*.*\0\0"_s);`

There is no point in creating a String to call makeString(). It is more efficient to pass an ASCIILiteral.

haruhisa-shin replied:
The results were different for String and ASCIILiteral.

> auto allFiles = makeString(allFilesText(), String("\0*.*\0\0", 6));

The length of `allFiles` is 15, and the filter field in the dialog is normal.

> auto allFiles = makeString(allFilesText(), "\0*.*\0\0"_s);

The length of `allFiles` is 9, and the filter field in the dialog is garbled.

This is just a guess, but I wonder if ASCIILiteral which contains null characters has some problem with the conversion in the middle.

haruhisa-shin replied:
In the first place, the filter with only "All Files" is not very useful, so I would like to delete this line itself.

haruhisa-shin replied:
The filter string setting has been removed.
The default is the same as selecting "All Files (*.*)".

fujii commented on Source/WebKit/UIProcess/win/PageClientImpl.cpp:
```diff
@@ -232,6 +240,70 @@ void PageClientImpl::preferencesDidChange()
+
+    ofn.lStructSize = sizeof(ofn);
+    ofn.hwndOwner = viewWindow;
+    String allFiles = makeString(allFilesText(), String("\0*.*\0\0", 6));
```
Can you use `"\0*.*\0\0"_s` here rather than `String("\0*.*\0\0", 6)`?

haruhisa-shin replied:
Ditto.

fujii commented on Source/WebKit/UIProcess/win/PageClientImpl.cpp:
```diff
@@ -232,6 +240,70 @@ void PageClientImpl::preferencesDidChange()
+    ofn.lpstrTitle = dialogTitleCharacters.data();
+    ofn.Flags = OFN_ENABLESIZING | OFN_FILEMUSTEXIST | OFN_PATHMUSTEXIST | OFN_EXPLORER;
+    if (isAllowMultipleFiles)
+        ofn.Flags = ofn.Flags | OFN_ALLOWMULTISELECT;
```
`ofn.Flags |= OFN_ALLOWMULTISELECT;`

haruhisa-shin replied:
Fixed at 19fcde7

fujii commented on Source/WebKit/UIProcess/win/PageClientImpl.cpp:
```diff
@@ -232,6 +240,70 @@ void PageClientImpl::preferencesDidChange()
+    ofn.hwndOwner = viewWindow;
+    String allFiles = makeString(allFilesText(), String("\0*.*\0\0", 6));
+
+    Vector<wchar_t> filterCharacters = allFiles.wideCharacters(); // Retain buffer long enough to make the GetOpenFileName call
```
Use `auto`.

haruhisa-shin replied:
Removed at 19fcde7

fujii commented on Source/WebKit/UIProcess/win/PageClientImpl.cpp:
```diff
@@ -232,6 +240,70 @@ void PageClientImpl::preferencesDidChange()
+    ofn.lpstrFile = fileBuf.data();
+    ofn.nMaxFile = fileBuf.size();
+    String dialogTitle = uploadFileText();
+    Vector<wchar_t> dialogTitleCharacters = dialogTitle.wideCharacters(); // Retain buffer long enough to make the GetOpenFileName call
```
Use `auto`. You are mixing `wchar_t` and `WCHAR`.  I guess WebKit is using `wchar_t` more than `WCHAR`

haruhisa-shin replied:
Fixed at 19fcde7

fujii commented on Source/WebKit/UIProcess/Inspector/win/WebInspectorUIProxyWin.cpp:
```diff
@@ -77,6 +78,63 @@ static InspectedWindowInfo getInspectedWindowInfo(HWND inspectedWindow, HWND par
+
+    wchar_t* messageBuffer = nullptr;
+    size_t size = FormatMessageW(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
+        nullptr, errorCode, MAKELANGID(LANG_NEUTRAL, SUBLANG_NEUTRAL), (LPTSTR)&messageBuffer, 0, nullptr);
```
Do you need `(LPTSTR)` cast here?
If so, use c++ style cast.

haruhisa-shin replied:
It was unnecessary. I'll remove cast.

haruhisa-shin replied:
Sorry, the cast was necessary. 
This function returns an `ERROR_INVALID_PARAMETER` error without a cast.
I'll use c++ style cast.

haruhisa-shin replied:
Fixed at https://github.com/WebKit/WebKit/commit/70ce9358458f9401040946ef3392f2ae55d6fa3e
`reinterpret_cast<LPWSTR>(&messageBuffer)`

fujii commented on Source/WebKit/UIProcess/Inspector/win/WebInspectorUIProxyWin.cpp:
```diff
@@ -77,6 +78,63 @@ static InspectedWindowInfo getInspectedWindowInfo(HWND inspectedWindow, HWND par
+            auto errMsg = systemErrorMessage(GetLastError());
+            if (errMsg.isEmpty())
+                errMsg = makeString("Error: writeToFile returns ", bytesWritten, ", contentLength = ", content.length());
+            MessageBox(0, errMsg.wideCharacters().data(), L"Export HAR", MB_OK | MB_ICONEXCLAMATION);
```
Do you have a problem if you pass `parentWindow` as the first argument?

haruhisa-shin replied:
Sure, nothing wrong.

haruhisa-shin replied:
Fixed at https://github.com/WebKit/WebKit/commit/70ce9358458f9401040946ef3392f2ae55d6fa3e

fujii commented on Source/WebKit/UIProcess/Inspector/win/WebInspectorUIProxyWin.cpp:
```diff
@@ -77,6 +78,63 @@ static InspectedWindowInfo getInspectedWindowInfo(HWND inspectedWindow, HWND par
+        String errMsg;
+        auto errCode = CommDlgExtendedError();
+        if (errCode) {
+            // FNERR_INVALIDFILENAME is treated specially becayse this is common error.
```
becayse → because
common → a common

haruhisa-shin replied:
Fixed at https://github.com/WebKit/WebKit/commit/70ce9358458f9401040946ef3392f2ae55d6fa3e

fujii commented on Source/WebKit/UIProcess/win/PageClientImpl.cpp:
```diff
@@ -42,6 +45,11 @@
+// When you call GetOpenFileName, if the size of the buffer is too small,
+// MSDN says that the first two bytes of the buffer contain the required size for the file selection, in bytes or characters
+// So we can assume the required size can't be more than the maximum value for a short.
+static const size_t maxFilePathsListSize = USHRT_MAX;
```
change `const` to `constexpr`, and move this into `handleRunOpenPanel`.

haruhisa-shin replied:
Fixed at https://github.com/WebKit/WebKit/commit/70ce9358458f9401040946ef3392f2ae55d6fa3e

fujii commented on Source/WebKit/UIProcess/win/PageClientImpl.cpp:
```diff
@@ -232,6 +240,63 @@ void PageClientImpl::preferencesDidChange()
+    OPENFILENAME ofn { };
+
+    // Need to zero out the first char of fileBuf so GetOpenFileName doesn't think it's an initialization string
+    fileBuf[0] = '\0';
```
L'\0'

haruhisa-shin replied:
Fixed at https://github.com/WebKit/WebKit/commit/70ce9358458f9401040946ef3392f2ae55d6fa3e

fujii commented on Source/WebKit/UIProcess/win/PageClientImpl.cpp:
```diff
@@ -232,6 +240,63 @@ void PageClientImpl::preferencesDidChange()
+    auto dialogTitle = uploadFileText();
+    auto dialogTitleCharacters = dialogTitle.wideCharacters(); // Retain buffer long enough to make the GetOpenFileName call
+    ofn.lpstrTitle = dialogTitleCharacters.data();
+    ofn.Flags = OFN_ENABLESIZING | OFN_FILEMUSTEXIST | OFN_PATHMUSTEXIST | OFN_EXPLORER;
```
If I read the document correctly, you don't need OFN_ENABLESIZING and OFN_EXPLORER here. Right?

haruhisa-shin replied:
https://learn.microsoft.com/en-us/windows/win32/api/commdlg/ns-commdlg-openfilenamea

OFN_ENABLESIZING is unnecessary as you say.
```
By default, the Explorer-style Open and Save As dialog boxes allow the dialog box to be resized regardless of whether this flag is set. 
```

If OFN_EXPLORER is not set, the old-style dialog that looks like Windows 3.1 will be shown.
```
OFN_ALLOWMULTISELECT 0x00000200
The File Name list box allows multiple selections. If you also set the OFN_EXPLORER flag, the dialog box uses the Explorer-style user interface; otherwise, it uses the old-style user interface.
```

Then I'll remove OFN_ENABLESIZING.

haruhisa-shin replied:
Fixed at https://github.com/WebKit/WebKit/commit/70ce9358458f9401040946ef3392f2ae55d6fa3e
Removed only OFN_ENABLESIZING.

fujii commented on Source/WebKit/UIProcess/win/PageClientImpl.cpp:
```diff
@@ -232,6 +240,63 @@ void PageClientImpl::preferencesDidChange()
+        return false;
+
+    bool isAllowMultipleFiles = parameters->allowMultipleFiles();
+    Vector<wchar_t> fileBuf(isAllowMultipleFiles ? maxFilePathsListSize : MAX_PATH);
```
You can't use `buf`. https://webkit.org/code-style-guidelines/#names-full-words
`fileBuf` → `fileBuffer`.

haruhisa-shin replied:
Fixed at https://github.com/WebKit/WebKit/commit/70ce9358458f9401040946ef3392f2ae55d6fa3e

fujii commented on Source/WebKit/UIProcess/win/PageClientImpl.cpp:
```diff
@@ -232,6 +240,63 @@ void PageClientImpl::preferencesDidChange()
+                    fileList.append(file);
+                files = nextFilePtr;
+                file = nextFile;
+            }
```
This loop should look like:
~~~c++
for (auto p = fileBuffer.data(); *p;) {
  auto length = wcslen(p);
  String filename(p, length);
  (...)
  p += length + 1;
}
~~~

haruhisa-shin replied:
When multiple files are selected, the directory and file names are given separately.
I found that the previous code did not work correctly.
I'll change the code based on the following information.
https://stackoverflow.com/questions/26317556/how-to-get-list-of-selected-files-when-using-getopenfilename-with-multiselect

haruhisa-shin replied:
Changed the around codes at https://github.com/WebKit/WebKit/commit/70ce9358458f9401040946ef3392f2ae55d6fa3e

fujii commented on Source/WebKit/UIProcess/win/PageClientImpl.cpp:
```diff
@@ -232,6 +235,69 @@ void PageClientImpl::preferencesDidChange()
+
+    OPENFILENAME ofn { };
+
+    // Need to zero out the first char of fileBuf so GetOpenFileName doesn't think it's an initialization string
```
fileBuf → fileBuffer

fujii commented on Source/WebKit/UIProcess/win/PageClientImpl.cpp:
```diff
@@ -232,6 +235,69 @@ void PageClientImpl::preferencesDidChange()
+            while (*p) {
+                length = wcslen(p);
+                String fileName(p, length);
+                fileList.append(FileSystem::pathByAppendingComponent(firstValue, fileName));
```
WTFMove(fileName)

fujii commented on Source/WebKit/UIProcess/win/PageClientImpl.cpp:
```diff
@@ -232,6 +235,69 @@ void PageClientImpl::preferencesDidChange()
+            }
+        } else
+            // If only one file is selected, one full path string is set in the buffer.
+            fileList.append(firstValue);
```
WTFMove(firstValue)

fujii commented on Source/WebKit/UIProcess/Inspector/win/WebInspectorUIProxyWin.cpp:
```diff
@@ -73,6 +74,63 @@ static InspectedWindowInfo getInspectedWindowInfo(HWND inspectedWindow, HWND par
+        }
+        FileSystem::closeFile(fd);
+    } else {
+        String errMsg;
```
Move errMsg into the following if block.
It might better to rename this variable. https://webkit.org/code-style-guidelines/#names-full-words
`message` or `text` or something?

haruhisa-shin committed:
[Win] WebInspector: enable HAR export and import
https://bugs.webkit.org/show_bug.cgi?id=253552

Reviewed by Fujii Hironori.

Add file open dialog implementations for both Export and Import.
And, canSave() now returns true so that the Export button can be pressed.

* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.h:
* Source/WebKit/UIProcess/Inspector/win/RemoteWebInspectorUIProxyWin.cpp:
(WebKit::RemoteWebInspectorUIProxy::platformSave):
* Source/WebKit/UIProcess/Inspector/win/WebInspectorUIProxyWin.cpp:
(WebKit::systemErrorMessage):
(WebKit::WebInspectorUIProxy::showSavePanelForSingleFile):
(WebKit::WebInspectorUIProxy::platformSave):
* Source/WebKit/UIProcess/win/PageClientImpl.cpp:
(WebKit::PageClientImpl::handleRunOpenPanel):
* Source/WebKit/UIProcess/win/PageClientImpl.h:
* Source/WebKit/WebProcess/Inspector/win/RemoteWebInspectorUIWin.cpp:
(WebKit::RemoteWebInspectorUI::canSave):
* Source/WebKit/WebProcess/Inspector/win/WebInspectorUIWin.cpp:
(WebKit::WebInspectorUI::canSave):

Canonical link: https://commits.webkit.org/261609@main
